module/zfs/zfs_znode.c

```diff
@@ -798,9 +799,10 @@ zfs_mknode(znode_t *dzp, vattr_t *vap, dmu_tx_t *tx, cred_t *cr,
 
 	if (S_ISDIR(vap->va_mode)) {
 		size = 2;		/* contents ("." and "..") */
-		links = (flag & (IS_ROOT_NODE | IS_XATTR)) ? 2 : 1;
+		links = 2;
```

Comment:
In most cases (`zpl_create()`, `zfs_mkdir()` etc.), after this function is called, the caller then uses `zfs_link_create()`, which incremented the link count (where we ran into the `I_LINKABLE` warning): so standard dirs were getting a link count of 1 here, then incremented to 2 in ` zfs_link_create()`. However the `zfs_make_xattrdir()` and `zfs_create_fs()` callers don't call ` zfs_link_create()`, so they pass the `IS_ROOT_NODE` or `IS_XATTR` flags into this function to indicate "set the proper link count".
This patch changes things around so this function always sets the correct link count, and `zfs_link_create()` doesn't increment the link count when it doesn't need to.